File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/8356b0a22079007d3a26cf866a88b12a.py
import sys
from collections import deque
from itertools import combinations

INF = 2147483647
MAXN = 1005
MAXK = 9

class Edge:
    def __init__(self, to, rev, cap):
        self.to = to
        self.rev = rev
        self.cap = cap

class MaxFlow:
    def __init__(self):
        self.src = 1002
        self.snk = 1003
        self.G = [[] for _ in range(MAXN)]
        self.dist = [-1] * MAXN
        self.seen = [0] * MAXN

    def clear(self):
        for i in range(MAXN):
            self.G[i].clear()

    def bfs(self, s):
        self.dist = [-1] * MAXN
        self.dist[s] = 0
        Q = deque([s])
        while Q:
            x = Q.popleft()
            for e in self.G[x]:
                if e.cap and self.dist[e.to] == -1:
                    self.dist[e.to] = self.dist[x] + 1
                    Q.append(e.to)

    def dfs(self, now, dest, fl):
        if now == dest:
            return fl
        for i in range(self.seen[now], len(self.G[now])):
            self.seen[now] = i
            ed = self.G[now][i]
            if ed.cap and self.dist[ed.to] == self.dist[now] + 1:
                f = self.dfs(ed.to, dest, min(fl, ed.cap))
                if f:
                    ed.cap -= f
                    self.G[ed.to][ed.rev].cap += f
                    return f
        return 0

    def dinic(self, s=None, t=None):
        if s is None: s = self.src
        if t is None: t = self.snk
        ret = 0
        while True:
            self.seen = [0] * MAXN
            self.bfs(s)
            if self.dist[t] == -1:
                return ret
            f = self.dfs(s, t, INF)
            while f:
                ret += f
                f = self.dfs(s, t, INF)

class Matrix:
    def __init__(self):
        self.dist = [[float('inf')] * 77 for _ in range(77)]

def read():
    return int(sys.stdin.readline().strip())

def qp(x, p, mod):
    ans = 1
    while p:
        if p & 1:
            ans = ans * x % mod
        x = x * x % mod
        p >>= 1
    return ans

def inv(x, mod):
    return qp(x, mod - 2, mod)

def init():
    global cc, id, T1
    cc = 0
    id = [0] * 256
    T1 = Matrix()
    for i in range(1 << k):
        cnt = bin(i).count('1')
        if cnt != x:
            continue
        cc += 1
        id[i] = cc
    for i in range(1 << k):
        cnt = bin(i).count('1')
        if cnt != x:
            continue
        v = i >> 1
        u = i & 1
        if not u:
            T1.dist[id[i]][id[v]] = 0
        else:
            for j in range(k):
                if v & (1 << j):
                    continue
                T1.dist[id[i]][id[v ^ (1 << j)]] = c[j + 1]

def build(pos):
    ret = T1
    for i in range(1 << k):
        if not id[i]:
            continue
        for j in range(1, q + 1):
            if p[j] <= pos and pos - p[j] < k:
                pp = pos - p[j]
                pp = k - 1 - pp
                if i >> (pp + 1) & 1:
                    continue
                for l in range(1 << k):
                    if not id[l]:
                        continue
                    if l >> pp & 1:
                        if ret.dist[id[i]][id[l]] != float('inf'):
                            ret.dist[id[i]][id[l]] += wp[j]
    return ret

def mul(x, y):
    ret = Matrix()
    for i in range(1, cc + 1):
        for j in range(1, cc + 1):
            for k in range(1, cc + 1):
                ret.dist[i][j] = min(ret.dist[i][j], x.dist[i][k] + y.dist[k][j])
    return ret

def multiply(X):
    global dp, dp2
    dp2 = [float('inf')] * 77
    for i in range(1, cc + 1):
        for j in range(1, cc + 1):
            dp2[j] = min(dp2[j], dp[i] + X.dist[i][j])
    dp = dp2[:]

def mqp(x, p):
    R = Matrix()
    for i in range(1, cc + 1):
        R.dist[i][i] = 0
    while p:
        if p & 1:
            R = mul(R, x)
        x = mul(x, x)
        p >>= 1
    return R

def normal(pos):
    for i in range(1, q + 1):
        if p[i] <= pos and pos - p[i] < k:
            return False
    return True

x = read()
k = read()
n = read()
q = read()
n += k - x
c = [0] * (MAXK + 1)
for i in range(1, k + 1):
    c[i] = read()
_init()
p = [0] * 30
wp = [0] * 30
for i in range(1, q + 1):
    p[i] = read()
    wp[i] = read()
dp = [float('inf')] * 77
dp[id[(1 << x) - 1]] = 0
now = k
while now < n:
    if not normal(now + 1):
        Tra = build(now + 1)
        multiply(Tra)
        now += 1
        continue
    Nxt = n + 1
    for i in range(1, q + 1):
        if p[i] > now:
            Nxt = min(Nxt, p[i])
    di = Nxt - now - 1
    Tra = mqp(T1, di)
    multiply(Tra)
    now = Nxt - 1

print(dp[id[(1 << x) - 1]])

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/8356b0a22079007d3a26cf866a88b12a.go
package main

import (
	"fmt"
	"math"
	"sort"
)

const (
	Maxn = 9
	big  = 20160116
	INF  = 2147483647
)

type LL int64

type edge struct {
	to, rev, cap int
}

type mat struct {
	dist [77][77]LL
}

var (
	x, k, n, q int
	tl          int
	id          [256]int
	T1          mat
	cc          int
	c           [Maxn]LL
	dp          [77]LL
	dp2         [77]LL
	p, wp       [30]int
)

func init() {
	for i := 0; i < 77; i++ {
		for j := 0; j < 77; j++ {
			T1.dist[i][j] = 1e18
		}
	}
}

func read() int {
	var x int
	fmt.Scan(&x)
	return x
}

func _init() {
	for i := 0; i < (1 << k); i++ {
		cnt := 0
		for j := 0; j < k; j++ {
			if i>>j&1 != 0 {
				cnt++
			}
		}
		if cnt != x {
			continue
		}
		id[i] = cc + 1
		cc++
	}
	for i := 0; i < (1 << k); i++ {
		cnt := 0
		for j := 0; j < k; j++ {
			if i>>j&1 != 0 {
				cnt++
			}
		}
		if cnt != x {
			continue
		}
		v := i >> 1
		u := i & 1
		if u == 0 {
			T1.dist[id[i]][id[v]] = 0
		} else {
			for j := 0; j < k; j++ {
				if v&(1<<j) != 0 {
					continue
				}
				T1.dist[id[i]][id[v^(1<<j)]] = c[j+1]
			}
		}
	}
}

func Build(pos int) mat {
	ret := T1
	for i := 0; i < (1 << k); i++ {
		if id[i] == 0 {
			continue
		}
		for j := 1; j <= q; j++ {
			if p[j] <= pos && pos-p[j] < k {
				pp := pos - p[j]
				pp = k - 1 - pp
				if i>>(pp+1)&1 != 0 {
					continue
				}
				for l := 0; l < (1 << k); l++ {
					if id[l] == 0 {
						continue
					}
					if l>>pp&1 != 0 {
						if ret.dist[id[i]][id[l]] != 1e18 {
							ret.dist[id[i]][id[l]] += LL(wp[j])
						}
					}
				}
			}
		}
	}
	return ret
}

func mul(x, y mat) mat {
	var ret mat
	for i := 1; i <= cc; i++ {
		for j := 1; j <= cc; j++ {
			for k := 1; k <= cc; k++ {
				ret.dist[i][j] = min(ret.dist[i][j], x.dist[i][k]+y.dist[k][j])
			}
		}
	}
	return ret
}

func Mul(X mat) {
	for i := 1; i <= cc; i++ {
		dp2[i] = 1e18
	}
	for i := 1; i <= cc; i++ {
		for j := 1; j <= cc; j++ {
			dp2[j] = min(dp2[j], dp[i]+X.dist[i][j])
		}
	}
	for i := 1; i <= cc; i++ {
		dp[i] = dp2[i]
	}
}

func mqp(x mat, p LL) mat {
	var R mat
	for i := 1; i <= cc; i++ {
		R.dist[i][i] = 0
	}
	for p > 0 {
		if p&1 != 0 {
			R = mul(R, x)
		}
		x = mul(x, x)
		p >>= 1
	}
	return R
}

func normal(pos int) bool {
	for i := 1; i <= q; i++ {
		if p[i] <= pos && pos-p[i] < k {
			return false
		}
	}
	return true
}

func min(a, b LL) LL {
	if a < b {
		return a
	}
	return b
}

func main() {
	x = read()
	k = read()
	n = read()
	q = read()
	n += k - x
	for i := 1; i <= k; i++ {
		c[i] = LL(read())
	}
	_init()

	for i := 1; i <= q; i++ {
		p[i] = read()
		wp[i] = read()
	}
	for i := 1; i <= cc; i++ {
		dp[i] = 1e18
	}
	dp[id[(1<<x)-1]] = 0
	now := k
	for now < n {
		if !normal(now + 1) {
			Tra := Build(now + 1)
			Mul(Tra)
			now++
			continue
		}
		Nxt := n + 1
		for i := 1; i <= q; i++ {
			if p[i] > now {
				Nxt = min(Nxt, p[i])
			}
		}
		di := Nxt - now - 1
		Tra := mqp(T1, LL(di))
		Mul(Tra)
		now = Nxt - 1
	}
	fmt.Println(dp[id[(1<<x)-1]])
}
